File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/TestDataFromCsvConverter.java
package org.skellig.teststep.processing.converter;

import de.siegmar.fastcsv.reader.CsvContainer;
import de.siegmar.fastcsv.reader.CsvReader;
import org.skellig.teststep.processing.exception.TestDataConversionException;

import java.net.URISyntaxException;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

class TestDataFromCsvConverter implements TestDataConverter {

    private static final String ROW_KEYWORD = "row";
    private static final String CSV_KEYWORD = "csv";
    private static final String FILE_KEYWORD = "file";

    private ClassLoader classLoader;

    TestDataFromCsvConverter(ClassLoader classLoader) {
        this.classLoader = classLoader;
    }

    public Object convert(Object value) {
        if (value instanceof Map) {
            Map<String, Object> valueAsMap = (Map<String, Object>) value;
            if (valueAsMap.containsKey(CSV_KEYWORD)) {
                Map<String, Object> csv = (Map<String, Object>) valueAsMap.get(CSV_KEYWORD);
                String csvFile = (String) csv.get(FILE_KEYWORD);

                value = readCsvFile(csvFile, getRowFilter(csv));
            }
        }
        return value;
    }

    private Predicate<Map<String, String>> getRowFilter(Map<String, Object> csvDetails) {
        Predicate<Map<String, String>> rowFilter;
        if (csvDetails.containsKey(ROW_KEYWORD)) {
            Map<String, Object> row = (Map<String, Object>) csvDetails.get(ROW_KEYWORD);
            rowFilter = (item) ->
                    row.entrySet().stream()
                            .allMatch(entry -> item.containsKey(entry.getKey()) && item.get(entry.getKey()).equals(entry.getValue()));
        } else {
            rowFilter = (row) -> true;
        }
        return rowFilter;
    }

    private List<Map<String, String>> readCsvFile(String fileName, Predicate<Map<String, String>> rowFilter) {
        List<Map<String, String>> result = new ArrayList<>();

        Path pathToFile = getPathToFile(fileName);
        if (Files.exists(pathToFile)) {
            readCsvContainer(pathToFile)
                    .ifPresent(csvContainer -> {
                        csvContainer.getRows()
                                .forEach(csvRow -> {
                                    Map<String, String> row = csvRow.getFieldMap().entrySet().stream()
                                            .collect(Collectors.toMap(
                                                    entry -> entry.getKey().trim(),
                                                    entry -> entry.getValue().trim()));
                                    if (rowFilter.test(row)) {
                                        result.add(row);
                                    }
                                });
                    });
        } else {
            throw new TestDataConversionException(String.format("File %s does not exist", fileName));
        }
        return result;
    }

    private Path getPathToFile(String fileName) {
        URL url = classLoader.getResource(fileName);
        if (url == null) {
            throw new TestDataConversionException(String.format("File '%s' was not found in resources", fileName));
        } else {
            try {
                return Paths.get(url.toURI());
            } catch (URISyntaxException e) {
                throw new TestDataConversionException(String.format("Failed to get path to '%s'", fileName), e);
            }
        }
    }

    private Optional<CsvContainer> readCsvContainer(Path pathToFile) {
        try {
            CsvReader csvReader = new CsvReader();
            csvReader.setContainsHeader(true);
            return Optional.ofNullable(csvReader.read(pathToFile, StandardCharsets.UTF_8));
        } catch (Exception e) {
            return Optional.empty();
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/TestDataFromCsvConverter.kt
package org.skellig.teststep.processing.converter

import de.siegmar.fastcsv.reader.CsvContainer
import de.siegmar.fastcsv.reader.CsvReader
import de.siegmar.fastcsv.reader.CsvRow
import org.skellig.teststep.processing.exception.TestDataConversionException
import org.skellig.teststep.processing.experiment.FunctionValueProcessor
import java.net.URISyntaxException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.function.Consumer
import java.util.function.Predicate
import java.util.stream.Collectors

class TestDataFromCsvConverter(val classLoader: ClassLoader) : FunctionValueProcessor {

    override fun execute(name: String, args: Array<Any?>): Any? {
        return if (args.isNotEmpty()) {
            val csvFile = args[0] as String
            val filter = if (args.size == 2 && args[1] is Map<*, *>) args[1] as Map<String, Any?> else null
            readCsvFile(csvFile, getRowFilter(filter))
        } else throw TestDataConversionException("Function `csv` can only accept 1 or 2 arguments. Found ${args.size}")
    }

    private fun getRowFilter(rowFilter: Map<String, Any?>?): Predicate<Map<String, String>> {
        return if (rowFilter != null) {
            Predicate { item: Map<String, String> ->
                rowFilter.entries.all { item.containsKey(it.key) && item[it.key] == it.value }
            }
        } else Predicate { true }
    }

    private fun readCsvFile(fileName: String?, rowFilter: Predicate<Map<String, String>>): List<Map<String, String>> {
        val result = mutableListOf<Map<String, String>>()
        val pathToFile = getPathToFile(fileName)
        if (Files.exists(pathToFile)) {
            readCsvContainer(pathToFile)
                .let { csvContainer ->
                    csvContainer?.rows?.forEach(Consumer { csvRow: CsvRow ->
                        val row = csvRow.fieldMap.entries.stream()
                            .collect(
                                Collectors.toMap<Map.Entry<String, String>, String, String>(
                                    { entry: Map.Entry<String, String> -> entry.key.trim { it <= ' ' } },
                                    { entry: Map.Entry<String, String> -> entry.value.trim { it <= ' ' } })
                            )
                        if (rowFilter.test(row)) {
                            result.add(row)
                        }
                    })
                }
        } else {
            throw TestDataConversionException(String.format("File %s does not exist", fileName))
        }
        return result
    }

    private fun getPathToFile(fileName: String?): Path {
        val url = classLoader.getResource(fileName)
        return if (url == null) {
            throw TestDataConversionException(String.format("File '%s' was not found in resources", fileName))
        } else {
            try {
                Paths.get(url.toURI())
            } catch (e: URISyntaxException) {
                throw TestDataConversionException(String.format("Failed to get path to '%s'", fileName), e)
            }
        }
    }

    private fun readCsvContainer(pathToFile: Path): CsvContainer? {
        return try {
            val csvReader = CsvReader()
            csvReader.setContainsHeader(true)
            csvReader.read(pathToFile, StandardCharsets.UTF_8)
        } catch (e: Exception) {
            null
        }
    }

    override fun getFunctionName(): String = "fromCsv"
}